driver/terraform.go

```diff
@@ -60,7 +60,8 @@ type Terraform struct {
 	logClient  bool
 	postApply  handler.Handler
 
-	inited bool
+	inited       bool
+	renderedOnce bool
```

Comment:
I was thinking about hashicat's API in regard to the ResolveEvent and how this sort of internal state field really shouldn't be necessary.

Right not the `Complete` field only returns as true if *that run* resulted in a completed template. That is Complete is false even when the template has been completely rendered (all fields filled in), if there was no new values. 

With the addition of `NoChange`, we now have a bit of a redundancy. Previously, as long as you ignored the initial runs that don't have the values, every run after the initial Complete=true indicated both that there was a change and that the template is complete. Now NoChange and Complete duplicate this information.

Basically I'm thinking we change Complete to always return true unless it is missing values. That way once it had completed once it would stay complete (unless it had hierarchical dependencies, like looping over all services and looking up service info on each). Then NoChange could be used to know if an update had occurred.

Sorry about this being a bit much for a PR comment, I started with a short comment and it sort of grew.

Reply:
No worries about it being long, the context is useful! And I agree with your sentiment that this is a hashicat detail that the caller doesn't need to manage. I wanted to work through a solution from the CTS side and it helped reveal some limitations fo the current hashicat API. While working on this, I was also tinkering on the hashicat side to see how a different API would work.

Yeah `Complete` ended up being a challenge to use only because now there are 2 threads attempting to render the same template. They end up racing each other -- UpdateTask and the background go routine waiting on the `watcher.Wait`.

+1 to your suggestions, adding a bit to it too to make this comment thread longer 😉 

1. Always return `Complete = true` unless it's missing values
2. `NoChange` to determine if an update had occurred
3. The [template's dirty state](https://github.com/hashicorp/hcat/blob/master/template.go#L139-L155) doesn't handle multiple calls to `resolver.Run` well. This works well with one caller, but CTS probably introduced a new use-case that caused a problem here.
    a. The first caller would receive that it's dirty and run the template, but the second concurrent call would not know that it's dirty and receive `NoChange`.
    b. I saw this happen with the main go routing waiting on `watcher.Wait` and the second caller ended up being `UpdateTask`. In this case, since UpdateTask is a CLI/API call, it is intended to be synchronous, but the end-user would see that there are no changes for the task to execute.
4. It might be worthwhile to cache the rendered template content and always return `Content` for the above two cases, `Complete = true` or `NoChange = true`. This may make it friendlier for multiple callers.

The `renderedOnce` tracking here was a weird to figure out due to the race from #3b. Since we want UpdateTask to execute synchronously, it exposed an edge case where hcat returns `NoChange` since the other caller received the `Complete` since only one could consume the `dirty` state. The `UpdateTask` thread would continue processing and attempt to run the task, but turns out the template file hasn't rendered yet (racing with the other go routine) and errors when using the expected file for Terraform.

We don't have merge this PR in preference for hcat changes, it was a useful exercise for me to work through to better understand the bug. Though I would like to merge the changes around the `WaitForEvent` test helper function.

Reply:
The only tricky bit with implementing the above is the Content caching. I don't think it would be good to just add it to tmpl.Execute(), so I'm giving it some time to percolate.

This could be merged as is if it will move things forward. Depends on whether it is better to have this working now so it isn't a blocker or wait until the hcat changes land to do it. I didn't mean for this to be a blocking review, more that it prompted the design thought and I wanted to communicate that.

Reply:
:nod: no strong feelings about the cache, it was an idea that came to mind around consistency. I think CTS would function just fine without a cache for `NoChange` since the template is written to disk here, and so if the file exists that would be the same as the cache.

I see. Thanks for clarifying it's non-blocking. We're lucky to be in close proximity so once hashicat changes are in movement, we can easily update CTS as well. Thanks @eikenb for challenging the current API and looking to improve it 🙂